File: src/test/java/com/github/njuro/jard/post/decorators/PostDecoratorTest.java
package com.github.njuro.jard.post.decorators;

import com.github.njuro.jard.post.Post;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.TestInstance;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@Slf4j
public abstract class PostDecoratorTest {

  protected static PostDecorator decorator;
  protected static Post post;

  @BeforeAll
  public void initAll() {
    post = Post.builder().build();
    decorator = initDecorator();
  }

  protected abstract PostDecorator initDecorator();

  protected void decoratePost(String body) {
    decoratePost(post, body);
  }

  protected void decoratePost(Post post, String body) {
    post.setBody(body);
    decorator.decorate(post);
    log.info(body + " -> " + post.getBody());
  }
}

File: src/test/kotlin/com/github/njuro/jard/post/decorators/PostDecoratorTest.kt
package com.github.njuro.jard.post.decorators

import com.github.njuro.jard.post.Post
import mu.KotlinLogging
import org.junit.jupiter.api.BeforeAll

private val log = KotlinLogging.logger { }

internal abstract class PostDecoratorTest {

    private val post: Post = Post.builder().build()
    private lateinit var decorator: PostDecorator

    @BeforeAll
    fun initAll() {
        decorator = initDecorator()
    }

    protected abstract fun initDecorator(): PostDecorator

    protected fun decorate(body: String) = decorate(post, body)

    protected fun decorate(post: Post, body: String): String {
        post.body = body
        decorator.decorate(post)
        log.info { "$body -> ${post.body}" }
        return post.body
    }

}